src/server/controllers/s3.js

```diff
@@ -12,6 +13,14 @@ const defaultConfig = {
 module.exports = function s3 (config) {
   config = Object.assign({}, defaultConfig, config)
 
+  // Use credentials to allow assumed roles to pass STS sessions in.
+  // If the user doesn't specify key and secret, the default credentials (process-env)
+  // will be used by S3 in calls below.
+  let credentials
+  if ('key' in config && 'secret' in config) {
+    credentials = new AWS.Credentials(config.key, config.secret, config.sessionToken)
```

Comment:
what is `config.sessionToken`? Are we expecting this to be set anywhere?

Reply:
config.sessionToken is an optional session token. It's unlikely that a user would specify this themselves, but I wire it in here in the event that they do, since it is possible and could be used by people making explicit STS calls. I'll explain what it is below, but here are the 3 use cases:

* User comes in with ONLY key/secret (conventional and NON-AWS workflow). AWS.Credentials is called with key, secret, undefined, which is an acceptable way to call AWS.Credentials, as it takes sessionToken as an optional parameter.
* User comes in with key/secret/sessionToken. AWS.Credentials is called with all 3 and provides credentials based on all 3.
* User comes in with none of these, and the default AWS SDK values are used. In that case, the variables: `AWS_SESSION_TOKEN` `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` are used.

Now. What is a session token? Well, a session token is what allows AWS to provide ephemeral users to its services. Rather than having a user with a static key and static secret, AWS provides "roles" for trusted AWS resources (including lambdas and EC2 instances). This is a form of federating user access, and in these cases, the key and secret are not static, they're generated from an AWS STS call. Since it can't simply look up the key/secret in a table of your users, it uses the session token to assist in the lookup.

Why do we want this? Roles are considered best practice in AWS when possible as you don't have to secure the storage of keys, nor do you have to roll keys or manage keys. Simply authorizing lambda or ec2 to invoke a role and assigning the role to a running service gives that service access to underlying AWS resources.

Reply:
alright, thank you for the clarification